java-checks/src/main/java/org/sonar/java/checks/synchronization/DoubleCheckedLockingCheck.java

```diff
@@ -0,0 +1,215 @@
+import static org.sonar.plugins.java.api.tree.Tree.Kind.NULL_LITERAL;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.SYNCHRONIZED_STATEMENT;
+
+@Rule(key = "S2168")
+public class DoubleCheckedLockingCheck extends IssuableSubscriptionVisitor {
+
+  private Deque<CheckedLocking> ifConditionSymbolStack = new LinkedList<>();
+  private Deque<Tree> synchronizedStmtStack = new LinkedList<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(IF_STATEMENT, SYNCHRONIZED_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (isIfFieldEqNullTree(tree)) {
+      IfStatementTree ifTree = (IfStatementTree) tree;
+      ifConditionSymbolStack.push(new CheckedLocking(ifTree));
+      visitIfStatement(ifTree);
+    }
+    if (tree.is(SYNCHRONIZED_STATEMENT)) {
+      synchronizedStmtStack.push(tree);
+    }
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (isIfFieldEqNullTree(tree)) {
+      ifConditionSymbolStack.pop();
+    }
+    if (tree.is(SYNCHRONIZED_STATEMENT)) {
+      synchronizedStmtStack.pop();
+    }
+  }
+
+  private static boolean isIfFieldEqNullTree(Tree tree) {
+    if (!tree.is(IF_STATEMENT)) {
+      return false;
+    }
+    IfStatementTree ifTree = (IfStatementTree) tree;
+    if (!ifTree.condition().is(EQUAL_TO)) {
+      return false;
+    }
+    BinaryExpressionTree eqRelation = (BinaryExpressionTree) ifTree.condition();
+    return (isField(eqRelation.leftOperand()) && eqRelation.rightOperand().is(NULL_LITERAL))
+      || (isField(eqRelation.rightOperand()) && eqRelation.leftOperand().is(NULL_LITERAL));
+  }
+
+  private void visitIfStatement(IfStatementTree ifTree) {
+    if (insideCriticalSection()) {
+      CheckedLocking parentIf = sameFieldAlreadyOnStack(ifConditionSymbolStack.peek());
+      if (parentIf.ifTree == ifTree) {
+        return;
+      }
+      if (thenStmtInitializeField(ifTree.thenStatement(), parentIf.field)
+        && !parentIf.field.isVolatile()
+        && !isAssignedAtomically(parentIf.field)) {
+        ImmutableList<JavaFileScannerContext.Location> flow = ImmutableList.of(new JavaFileScannerContext.Location("Double-checked locking", ifTree.ifKeyword()));
+        reportIssue(parentIf.ifTree.ifKeyword(), "Remove this dangerous instance of double-checked locking.", flow, null);
+      }
+    }
+  }
+
+  private boolean insideCriticalSection() {
+    return !synchronizedStmtStack.isEmpty();
+  }
+
+  /**
+   * Returns if statement which has the same condition as nestedIf , or nestedIf if there is no such other
+   * if statement on the stack
+   */
+  private CheckedLocking sameFieldAlreadyOnStack(CheckedLocking nestedIf) {
+    Iterator<CheckedLocking> stackIterator = ifConditionSymbolStack.descendingIterator();
+    return Iterators.tryFind(stackIterator, parentIf -> parentIf.field == nestedIf.field)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)